noxfile.py

```diff
@@ -0,0 +1,279 @@
+]
+
+
+def set_environment(session: "nox.session", cc: str, check: bool = True) -> None:
+    if check and (shutil.which(cc) is None):
+        session.env["CC_REFERENCE"] = cc
+        cc = "gcc"
+        cxx = "g++"
+        gcov = "gcov"
+    else:
+        cxx = cc.replace("clang", "clang++").replace("gcc", "g++")
+        if cc.startswith("clang"):
+            gcov = cc.replace("clang", "llvm-cov") + " gcov"
+        else:
+            gcov = cc.replace("gcc", "gcov")
+    session.env["GCOVR_TEST_SUITE"] = "1"
+    session.env["CC"] = cc
+    session.env["CFLAGS"] = "--this_flag_does_not_exist"
+    session.env["CXX"] = cxx
+    session.env["CXXFLAGS"] = "--this_flag_does_not_exist"
+    session.env["GCOV"] = gcov
+
+
+@nox.session
+def qa(session: "nox.session") -> None:
+    for session_id in nox.options.sessions:
+        session.log("Notify session {}".format(session_id))
+        session.notify(session_id)
+
+
+@nox.session
+def lint(session: "nox.session") -> None:
+    session.install("flake8")
+    if platform.python_implementation() == "CPython":
+        session.install("black")
+    if session.posargs:
+        args = session.posargs
+    else:
+        args = DEFAULT_TEST_DIRECTORIES
+    session.run("flake8", *args)
+
+    if platform.python_implementation() == "CPython":
+        if session.posargs:
+            session.run("python", "-m", "black", *session.posargs)
+        else:
+            session.run(
+                "python", "-m", "black", "--diff", "--check", *BLACK_CONFORM_FILES
+            )
+            session.run("python", "-m", "black", "--diff", *DEFAULT_TEST_DIRECTORIES)
+    else:
+        session.log(
+            f"Skip black because of platform {platform.python_implementation()}."
+        )
+
+
+@nox.session
+def black(session: "nox.session") -> None:
+    session.install("black")
+    if session.posargs:
+        args = session.posargs
+    else:
+        raise RuntimeError("Please add the files to format as argument.")
+    session.run("python", "-m", "black", *args)
+
+
+@nox.session
+def doc(session: "nox.session") -> None:
+    session.install(
+        "sphinx",
+        "sphinx_rtd_theme",
+        "sphinxcontrib-autoprogram==0.1.5 ; python_version=='3.6'",
+        "sphinxcontrib-autoprogram>=0.1.5 ; python_version>='3.7'",
+    )
+    session.install("-e", ".")
+    session.run("bash", "-c", "cd doc && make html O=-W", external=True)
+
+
+@nox.session
+def tests(session: "nox.session") -> None:
+    session_id = "tests_version({})".format(GCC_VERSION2USE)
+    session.log("Notify session {}".format(session_id))
+    session.notify(session_id)
+
+
+@nox.session
+def tests_all_versions(session: "nox.session") -> None:
+    for version in GCC_VERSIONS:
+        session_id = "tests_version({})".format(version)
+        session.log("Notify session {}".format(session_id))
+        session.notify(session_id)
+
+
+@nox.session
+@nox.parametrize("version", [nox.param(v, id=v) for v in GCC_VERSIONS])
```

Comment:
This Noxfile uses the concept of a “version” a lot. I think talking about a `compiler` might be less confusing than a `version`, since the version could also refer to the gcovr or Python version.

Reply:
Ok, I'll rename this.